src/kinsol/kinsol.c

```diff
@@ -2865,8 +2893,30 @@ static int KINFP(KINMem kin_mem)
     /* compute new solution */
     if (kin_mem->kin_m_aa == 0 || kin_mem->kin_nni - 1 < kin_mem->kin_delay_aa)
     {
-      if (kin_mem->kin_damping)
+      if (kin_mem->kin_damping || kin_mem->kin_damping_fn)
       {
+        if (kin_mem->kin_damping_fn)
+        {
+          retval = kin_mem->kin_damping_fn(kin_mem->kin_nni, kin_mem->kin_uu,
+                                           kin_mem->kin_fval, NULL, 0,
+                                           kin_mem->kin_user_data,
+                                           &(kin_mem->kin_beta));
+          if (retval)
+          {
+            KINProcessError(kin_mem, KIN_DAMPING_FN_ERR, __LINE__, __func__,
+                            __FILE__, "The damping function failed.");
+            ret = KIN_DAMPING_FN_ERR;
+            break;
+          }
+          if (kin_mem->kin_beta <= ZERO || kin_mem->kin_beta > ONE)
+          {
+            KINProcessError(kin_mem, KIN_DAMPING_FN_ERR, __LINE__, __func__,
+                            __FILE__, "The damping parameter is outside of the range (0, 1].");
```

Comment:
Is the restriction beta<=1 needed?

Reply:
Yes, valid beta values need to be in the range (0, 1]